### ISSUE TITLE
Bug: [no-unnecessary-condition] Unnecessary conditional when trying to use coalesce operator on null | undefined

### ISSUE THREAD

**Issue by darkbasic** [typescript-eslint/typescript-eslint#10055]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.tsx&code=GYVwdgxgLglg9mABAWwJ4DFwQBQEMBOA5gFyJggA2FiAPouACYCmwMYTDAlKYy2x4gDeAKESJ8TKCHxIChRAH4F9MM1bsGwgL5A&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRgA2uGyRE0aB2iQANKuyYDaofHgcA7gGFJyfEJYAZDh2I3ps%2BU1QZ8cRMYQAL7GALqqIUFAA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
function myFunc(arg: null | undefined): undefined {
  return arg ?? undefined
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-condition": [
    "error",
      {
      "allowConstantLoopConditions": true
      }
    ]
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

`arg ?? undefined` limits the union type `null | undefined` to just `undefined` so it's not unnecessary.

### Actual Result

`arg ?? undefined` errors out like if it didn't restrict the union type a subset of its values.

**Comment by kirkwaiblinger** [comment]
The rule is trying to point out that you could write this like 
```ts
function myFunc(arg: null | undefined): undefined {
  return undefined
}
```

it reports both always-true and always-false conditions as unnecessary, too, like
```ts
declare const alwaysTrue: true;
declare const alwaysFalse: false;
// reported by the rule
if (alwaysTrue) {
}
// reported by the rule
if (alwaysFalse) {
}
```

**Comment by darkbasic** [comment]
Makes sense. My bad but I'm in the process of refactoring a big codebase with eslint-config-love and I'm a bit overflowed by over-reaching rules right now :)

**Comment by darkbasic** [comment]
@kirkwaiblinger sorry for reusing the same issue but I've also noticed the following:

```ts
let api: Axios

export function useApi(options: AxiosRequestConfig = {}): Axios {
  // Unnecessary conditional, the types have no overlap.eslint[@typescript-eslint/no-unnecessary-condition]
  if (api == null) {
    createApi(options)
  }
  return api
}
```

Typescript should be able to notice that no value has been assigned for certain to api and thus it can be undefined, right?

**Comment by kirkwaiblinger** [comment]
Ah - this is a pet peeve of mine - see https://github.com/typescript-eslint/typescript-eslint/issues/9565.

In short - no. `let x: Axios` means "`x` will be defined by the time I use it in a closure" and `let x: Axios | undefined` means "`x` may or may not be defined by the time it's accessed". This is a textbook case of where you _must_ use `let x: Axios | undefined` in order to get correct types.

**Comment by kirkwaiblinger** [comment]
For future reference, these sort of general usage/support questions are best asked in our discord server 🙂

**Comment by kirkwaiblinger** [comment]
> In short - no. 

Or, rather - it should complain, but it doesn't! 😆 

(note that I have yet to file the TS issue that I intended to file)

**Comment by darkbasic** [comment]
> it should complain, but it doesn't!

Oh that explains why they recently enabled init-declarations in eslint-love: if you're forced to declared a variable as undefined you will be forced to add undefined to the union as well. I sometimes tend to give Typescript more credit than it deserves: I would have bet that it was capable of determining if a variable has been assigned or not by the time it's being accessed :(

> note that I have yet to file the TS issue that I intended to file

Too bad, I would have added it to my (very long) list of Typescript issues I follow :)

> For future reference, these sort of general usage/support questions are best asked in our discord server

**Comment by kirkwaiblinger** [comment]
@darkbasic Opened it! https://github.com/microsoft/TypeScript/issues/60064 Thanks for the nudge 🙂
